tracing/tempo/cache.go

```diff
@@ -0,0 +1,181 @@
+package tempo
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+	"unsafe"
+
+	kialiConfig "github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+	"github.com/kiali/kiali/models"
+	"github.com/kiali/kiali/tracing/jaeger/model"
+)
+
+type (
+	queryResult struct {
+		queryTime time.Time // Expiration time
+		inResult  *model.TracingResponse
+	}
+
+	traceResult struct {
+		queryTime time.Time // Expiration time
+		inResult  *model.TracingSingleTrace
+	}
+
+	TempoCache interface {
+		GetAppTracesHTTP(service string, q models.TracingQuery) (bool, *model.TracingResponse)
+		GetTraceDetailHTTP(traceID string) (bool, *model.TracingSingleTrace)
+		SetAppTracesHTTP(service string, q models.TracingQuery, response *model.TracingResponse)
+		SetTraceDetailHTTP(traceID string, response *model.TracingSingleTrace)
+	}
+
+	tempoCacheImpl struct {
+		cacheExpiration time.Duration
+		// Cache by service name + query (Converted to string)
+		cacheAppTraces map[string]*queryResult
+		// Cache individual traces
+		cacheTraceDetails map[string]*traceResult
+		appTracesLock     sync.RWMutex
+		traceDetailsLock  sync.RWMutex
+		// stats
+		hits          int
+		totalRequests int
+	}
+)
+
+func NewTempoCache() TempoCache {
+	kConfig := kialiConfig.Get()
+
+	cacheExpiration := time.Duration(kConfig.ExternalServices.Tracing.TempoConfig.CacheExpiration) * time.Second
+
+	tempoCacheImpl := tempoCacheImpl{
+		cacheExpiration:   cacheExpiration,
+		cacheAppTraces:    make(map[string]*queryResult),
+		cacheTraceDetails: make(map[string]*traceResult),
```

Comment:
Have you seen the [ExpirationStore](https://github.com/kiali/kiali/blob/master/store/expiration_store.go). It already handles expiring keys after a certain time period and it's threadsafe. The KialiCache uses it for a few different things: https://github.com/kiali/kiali/blob/31bcc777ea265c4ec971f25608ecbbf52c291ca8/kubernetes/cache/cache.go#L146 It doesn't show you stats but you can add that if needed.